File: daily/Linked List/147-Insertion-sort-list.go
package daily

func insertionSortList(head *ListNode) *ListNode {
	dummy := &ListNode{Val: 0}
	curr := head

	for curr != nil {
		next_node := curr.Next
		prev := dummy

		for prev.Next != nil && prev.Next.Val < curr.Val {
			prev = prev.Next
		}

		curr.Next = prev.Next
		prev.Next = curr

		curr = next_node
	}

	return dummy.Next
}

File: daily/Linked List/147-Insertion-sort-list.py
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)   # sorted list starts empty
        curr = head

        while curr:
            next_node = curr.next   # save next before relinking
            prev = dummy

            # Find the correct position in sorted portion
            while prev.next and prev.next.val < curr.val:
                prev = prev.next

            # Splice curr into sorted portion
            curr.next = prev.next
            prev.next = curr

            curr = next_node  # move to next unsorted node

        return dummy.next